### DISCUSSION TITLE
What does write_texture()'s "size" parameter do?

### DISCUSSION THREAD

**Discussion by ChevyRay** [gfx-rs/wgpu#2497] [Q&A] [answered]
I'm building a wrapper around wgpu for a project and ran into something I was curious about:

So [Device::create_texture()](https://docs.rs/wgpu/latest/wgpu/struct.Device.html#method.create_texture) creates a texture, and the descriptor requires you to provide a `size`. But [Queue::write_texture()](https://docs.rs/wgpu/latest/wgpu/struct.Queue.html#method.write_texture), which is used for writing data into these created textures, also takes a `size` parameter.

If `Device` already knows the size of the texture, why does `Queue` require me to provide this? Does it always have to match the size the texture was created with? What happens if it's smaller than the original size? What if it's larger than the original size?

I can't find any of this information in the documentation, so I'd love if someone who knew could share.

**Comment by cwfitzgerald** [comment, marked as answer]
It's so that you can write to only part of the texture! You can use the [origin field of the image copy](https://docs.rs/wgpu/latest/wgpu/struct.ImageCopyTextureBase.html#structfield.origin) and the size, to form a rectangle that you want to write data to.

If either size or origin cause the box to overrun the texture, there will be a validation error.

We'd love for this documentation to be improved, PRs accepted :)
